packages/connect-wallet-modal/src/components/Ledger/hooks.ts

```diff
@@ -0,0 +1,135 @@
+import { useCallback, useContext, useEffect, useState } from 'react';
+import { fetchBalance } from 'wagmi/actions';
+import { LedgerContext } from './LedgerContext';
+import { AccountRecord, AccountsStorage } from './types';
+import {
+  getDerivationPathsForPage,
+  getFirstIndexOnPage,
+  isDeviceBusy,
+} from './helpers';
+
+export const useLedgerContext = () => useContext(LedgerContext);
+
+type LedgerAccountsHookParams = {
+  accountsStorage?: AccountsStorage;
+  derivationPathTemplate: string;
+  currentPage: number;
+  accountsPerPage: number;
+};
+
+export const useLedgerAccounts = ({
+  accountsStorage,
+  derivationPathTemplate,
+  currentPage,
+  accountsPerPage,
+}: LedgerAccountsHookParams) => {
+  const { ledgerAppEth, connectTransport, transport, setError } =
+    useLedgerContext();
+
+  const [accounts, setAccounts] = useState<Record<string, AccountRecord>>({});
+
+  const getAccountRecords = useCallback(
+    async (params: LedgerAccountsHookParams) => {
+      const eth = ledgerAppEth?.current;
+      // getAccountRecords can be called several times during renders in a short amount of time.
+      // Multiple parallel requests to Ledger results into an error, so there is a check if a device is busy or not.
+      if (isDeviceBusy(transport) || !eth) return;
+
+      const derivationPaths = getDerivationPathsForPage(
+        params.derivationPathTemplate,
+        params.currentPage,
+        params.accountsPerPage,
+      );
+
+      const accountRecords: AccountRecord[] = [];
+
+      const getAndPushAccount = async (index: string, path: string) => {
+        const { address } = await eth.getAddress(path);
+        const { formatted, symbol } = await fetchBalance({ address } as {
+          address: `0x${string}`;
+        });
+        accountRecords.push({
+          index,
+          path,
+          pathTemplate: derivationPathTemplate,
+          address,
+          balance: Number(formatted),
+          token: symbol,
+        });
+      };
+
+      // iterate over derivation paths for the given page
+      for await (const [index, path] of Object.entries(derivationPaths)) {
```

Comment:
I'm almost sure `await` is not needed here
```suggestion
      for (const [index, path] of Object.entries(derivationPaths)) {
```

Reply:
if you want to do it in parallel, you need to do `Promise.all(derivationPaths.map)`, `for await` is used for async iterators (e.g. `Object.entries` would be returning promises here, not `[index, path]` tuples)

Reply:
Yes, you are right, it is a mistake.
Parallel requests are not going to work because ledger can't handle them.